src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,365 @@
+                return Err(EngineError::Engine(ErrorEnum::Invalid,
+                                               "data larger than region_size".into()));
             }
+
+            try!(f.seek(SeekFrom::Start(offset)));
+            try!(f.write_all(&hdr_buf));
+            try!(f.write_all(data));
+            try!(f.flush());
+
+            Ok(())
         };
-        Ok(ownership)
+
+        let older_region = self.older();
+
+        // Save to primary and backup regions
+        // TODO: Should we ignore errors?
+        try!(save_region(older_region));
+        try!(save_region(older_region + 2));
+
+        self.mdas[older_region].last_updated = *time;
+        self.mdas[older_region].used = data.len() as u64;
+        self.latest_json = Some(data.to_vec());
+
+        Ok(())
+    }
+
+    pub fn older(&mut self) -> usize {
+        match self.mdas[0].last_updated.cmp(&self.mdas[1].last_updated) {
+            Ordering::Less => 1,
+            Ordering::Greater => 0,
+            Ordering::Equal => 1,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct MDAHeader {
+    pub last_updated: Timespec,
+    pub used: u64,
+}
+
+impl MDAHeader {
+    pub fn new() -> MDAHeader {
+        MDAHeader {
+            last_updated: Timespec::new(0, 0),
+            used: 0,
+        }
+    }
+
+    pub fn from_buf(buf: &[u8; MDA_REGION_HDR_SIZE]) -> EngineResult<(MDAHeader, u32)> {
+
+        let hdr_crc = LittleEndian::read_u32(&buf[..4]);
+
+        if hdr_crc != crc32::checksum_ieee(&buf[4..]) {
+            return Err(EngineError::Engine(ErrorEnum::Invalid, "MDA region header CRC".into()));
+        }
+
+        Ok((MDAHeader {
+                used: LittleEndian::read_u64(&buf[8..16]),
+                last_updated: Timespec::new(LittleEndian::read_i64(&buf[16..24]),
+                                            LittleEndian::read_i32(&buf[24..28])),
+            },
+            LittleEndian::read_u32(&buf[4..8])))
+    }
+
+    pub fn to_buf(data: &[u8], timestamp: &Timespec) -> [u8; MDA_REGION_HDR_SIZE] {
+        let mut buf = [0u8; MDA_REGION_HDR_SIZE];
+
+        LittleEndian::write_u32(&mut buf[4..8], crc32::checksum_ieee(data));
+        LittleEndian::write_u64(&mut buf[8..16], data.len() as u64);
+        LittleEndian::write_u64(&mut buf[16..24], timestamp.sec as u64);
+        LittleEndian::write_u32(&mut buf[24..28], timestamp.nsec as u32);
+
+        buf
     }
 }
 
 /// Validate MDA size
 /// Return None if MDA size is fine, otherwise a message.
-pub fn validate_mda_size(size: Sectors) -> Option<String> {
-    if *size % NUM_MDA_COPIES != 0 {
-        let error_message = format!("MDA size {} is not divisible by number of copies \
-                                    required {}",
-                                    *size,
-                                    NUM_MDA_COPIES);
-        return Some(error_message);
+pub fn validate_mda_size(size: Sectors) -> EngineResult<()> {
```

Comment:
How so? The design doc defines two conditions for a valid mda size: it must be above a certain minimum and it needs to be divisible by 4. I am of the opinion that if an invalid value is passed to `BlockDev::initialize()` or read from the disk in `StaticHeader::from_buf`, something is seriously wrong and we should return an `EngineError`. This allows the rest of the code to assume the mda size is valid, which lets us for example calculate mda region starting offsets.